### PR TITLE
Blog post: A high-security API management infrastructure using Apache…

### PR DESCRIPTION
**Pull request by Yang-Xie-OSS**
… Camel

Here is a blog post that introduces a high-security API management infrastructure using Camel and Keycloak.

### CODE REVIEW

**Comment by Yang-Xie-OSS** [comment]
I noticed that I should create a JIRA before the pull request? But I find there is no JIRA about creating a blog post in [1].
So I haven't created it yet. If I must create it, please tell me.
[1] https://issues.apache.org/jira/browse/CAMEL-17326?jql=project%20%3D%20CAMEL

**Comment by Yang-Xie-OSS** [comment]
Hello @zregvart,
This is the blog post you invited. I'm sorry it's a bit late because we should do some examination procedures in our organization before publishing.

File: content/blog/2021/12/api-management-infra/index.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+```
+#### Effect
+
+As a result of implementing token introspection, the API request with an inactive access token will be denied by security enhanced API management infrastructure (with a 401 response). It means the drawback 1 is overcome.
+
+### Scope check
+
+{{< image "Scope-check.png" "Scope check" >}}
+
+Scope is a mechanism for limiting an application�fs access to an API. It is defined in [RFC6749](https://datatracker.ietf.org/doc/html/rfc6749), the specification document of OAuth 2.0.
```
Comment (tadayosi) [inline content/blog/2021/12/api-management-infra/index.md:130]: There is some non-ascii character here: `applicationfs`

File: content/blog/2021/12/api-management-infra/index.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+After receiving the token request with the scopes specified in query parameters, authorization server issues an access token with a scope member. The value of the scope member is a list of space-delimited strings (the scopes granted to the access token).
+
+The following is an example access token with a scope member.
+
+```json
+{
+  "iss": "https://example.hitachi.com/",
+  "aud": "https://app1.hitachi.com/",
+  "sub": �gjdoe",
+  "scope": �gread write dolphin",
```
Comment (tadayosi) [inline content/blog/2021/12/api-management-infra/index.md:141]: There are some non-ascii characters here as well:
```
  "sub": gjdoe",
  "scope": gread write dolphin",
```

File: content/blog/2021/12/api-management-infra/index.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+The access token used in OAuth MTLS should present the client certificate of the client application using a cnf.x5t#S256 member. The value of the cnf.x5t#S256 member is the the SHA-256 thumbprint of the client certificate.
+
+The following is an example access token that has a cnf.x5t#S256 member.
+
+```json
+{
+  "iss": "https://example.hitachi.com",
+  "aud": "https://app1.hitachi.com"
+  "sub": �gjdoe",
```
Comment (tadayosi) [inline content/blog/2021/12/api-management-infra/index.md:190]: Here as well:
```
  "sub": gjdoe",
```

**Review by tadayosi — CHANGES_REQUESTED** [review]

File: content/blog/2021/12/api-management-infra/index.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+}
+```
+
+#### Support in Keycloak
+
+For supporting token introspection, Keycloak provides an introspection endpoint to receive the introspection request.
+
+After receiving the introspection request, Keycloak inspects the access token with several steps including validate the session linked with the access token.
+
+Session is a data structure used in Keycloak for storing user�fs login information. Access token is generated from session and every access token is linked with one session. Access token and the linked session have the same value of their validities. Therefore, if the linked session is validated to invalid, the access token also will be validated to invalid even if its expiration time hasn�ft been reached.
```
Comment (tadayosi) [inline content/blog/2021/12/api-management-infra/index.md:99]: Non-ascii character here: `userfs`

File: content/blog/2021/12/api-management-infra/index.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+As shown in the picture below, the API management infrastructure can perform reverse proxy, protocol conversion, data conversion, mash-up, flow control, API documentation publishing and metrics. Besides, it also can perform simple API authorization by token issuance & management that is provided by Keycloak.
+
+{{< image "API-management-infrastructure.png" "API management infrastructure" >}}
+
+## Drawbacks of security
+
+Although the existing API management infrastructure has taken a security measure as token issuance & management, there are also three drawbacks of its security:
+
+1. Inadequate token validation. 
+1. No API access management for each API.
```
Comment (oscerd) [inline content/blog/2021/12/api-management-infra/index.md:30]: I think here we have point 2 and point 3 not 1.

**Review by oscerd — APPROVED** [review]
just a minor finding.

**Comment by Yang-Xie-OSS** [comment]
@tadayosi 
Thanks for the review. I've fixed them. Please check.

**Comment by Yang-Xie-OSS** [comment]
@oscerd 
Thanks for the review. I've fixed it. Please check.

**Comment by Yang-Xie-OSS** [comment]
@oscerd 
Thanks for the review. I've fixed it. Please check.

File: content/blog/2021/12/api-management-infra/index.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
+---
+title: "A high-security API management infrastructure using Apache Camel"
+date: 2021-12-13
```
Comment (davsclaus) [inline content/blog/2021/12/api-management-infra/index.md:3]: Update date to today

**Comment by Yang-Xie-OSS** [comment]
@davsclaus 
Thanks for the review. I've updated the date. Please check.

**Review by davsclaus — APPROVED** [review]

**Comment by zregvart** [comment]
@Yang-Xie-OSS thanks for the blog post!

The preview is here https://pr-715--camel.netlify.app/blog/2021/12/api-management-infra/

File: content/blog/2021/12/api-management-infra/index.md (hunk 5 above)
Comment (zregvart) [reply content/blog/2021/12/api-management-infra/index.md]: @oscerd can you elaborate, the point of this project as I understood it to add additional token validation, i.e. test for invalidated tokens. Look at the `Drawback 1: Inadequate token validation` section

**Review by zregvart — APPROVED** [review]
Thanks! This looks good to me, I'll leave this for a bit for any other folk to add comments before we merge this.

File: content/blog/2021/12/api-management-infra/index.md (hunk 5 above)
Comment (oscerd) [reply content/blog/2021/12/api-management-infra/index.md]: I meant say it should .1,.2 and .3 and not .1,.1 and .1

**Pull request merged**